src/cmd/mod.rs

```diff
@@ -107,36 +108,71 @@ pub async fn parse_response<T: 'static + DeserializeOwned>(res: reqwest::Respons
     let status = res.status();
     let etag = res.headers().get(http::header::ETAG).cloned();
     if status == StatusCode::NOT_MODIFIED {
-        Ok(Response::new(None, etag))
+        return Ok(Response::new(None, etag));
+    }
+    let is_json = res
+        .headers()
+        .get(http::header::CONTENT_TYPE)
+        .is_some_and(|h| h.as_bytes() == b"application/json");
+    let empty: Box<dyn Any + Send + Sync> = Box::new(());
+
+    const MAX_RESPONSE_SIZE: usize = 5_000_000;
+    let body = {
+        let mut body = Vec::new();
+        let mut chunks = res.bytes_stream();
+        loop {
+            match chunks.next().await {
+                Some(Ok(chunk)) => {
+                    if chunk.len() + body.len() > MAX_RESPONSE_SIZE {
+                        return Err(ClientError::ResponseTooLarge);
+                    }
+                    body.extend_from_slice(&chunk);
+                }
+                None => break body,
+                Some(Err(error)) => {
+                    return Err(ProtocolError {
+                        status,
+                        raw: String::new(),
+                        source: error.into(),
+                    }
+                    .into())
+                }
+            }
+        }
+    };
+
+    if !is_json {
+        Err(ProtocolError {
+            status,
+            raw: String::from_utf8(body).unwrap_or_default(),
+            source: anyhow::anyhow!("Non-JSON {status} response"),
+        }
+        .into())
+    } else if !status.is_success() {
+        match serde_json::from_slice(&body) {
+            Ok(api_error_body) => Err(ApiError {
+                status,
+                body: api_error_body,
+            }
+            .into()),
+            Err(error) => Err(ProtocolError {
+                status,
+                raw: String::from_utf8(body).unwrap_or_default(),
+                source: error.into(),
+            }
+            .into()),
+        }
+    } else if let Ok(empty) = empty.downcast::<T>() {
```

Comment:
Now that we're reading the body regardless, special-casing this doesn't seem as useful. This was originally meant to just be an optimization to skip reading and deserializing for `()`, right?

Reply:
It's not an optimization. An empty vector doesn't deserialize to `()`, so we just ignore the body if `T=()`. This gives us backward compatibility if we ever want to add a body in a future version.

Reply:
...which would actually deserialize to `()`. I see what you mean. This code is older than the backend code you are referring to. I don't really want to change behavior for `T=()` in this PR. If we want to not skip the body in the future we can consider that separately (so we can think about whether we want to sacrifice the backward compatibility aspect).